### ISSUE TITLE
Garbage collection awareness

### ISSUE THREAD

**Issue by obo20** [ipfs/ipfs-cluster#889]
When a node is in a garbage collection state, it would be helpful to have cluster recognize that. 

This way when you attempt to pin something, cluster doesn't try and allocate content to a node that's going through garbage collection and error out.

This could be possibly done by triggering garbage collection from cluster itself, but it doesn't necessarily have to be. I just don't want to have to worry about cluster producing faulty errors due to how IPFS garbage collection "stops the world" (preventing adds / removes)

There's definitely some things to consider here in terms of how this information gets propagated throughout the cluster, but this could be an incredibly helpful feature until IPFS garbage collection becomes more stable / asynchronous.

**Comment by lanzafame** [comment]
Cluster already performs hourly `recover` operations to handle when pins get into weird errors. I would be interested to see what errors cluster would produce in this scenario, @obo20 do you have any logs for this? 

If an hour is too long you can lower the `PinRecoverInterval` config value to what will work best for your setup.

**Comment by lanzafame** [comment]
> would it not be desirable to just have cluster know that a particular node is GC-ing and avoid trying to pin to them during that time?

Not really. It is added complexity for very little benefit. Theoretically, it would be ideal if we didn't rely on `recover` but unless go-ipfs provides an event stream for cluster to listen to for when it starts/stops GCing, internally cluster is going to be polling go-ipfs either with a 'are you gcing?' request or a 'pin this for me' request until the file pins.

**Comment by hsanjuan** [comment]
There was the idea of using cluster to orchestrate when GC happens (so, disabling auto-gc on IPFS). As a way to ensure that GC does not happen at the same time everywhere. And while cluster has triggered that GC, then it could also keep the node at bay by not allocating or adding more content to it while GC lasts.

But the first part is the most interesting as it would need a coordination mechanism between peers.

**Comment by obo20** [comment]
@hsanjuan is describing what I had in mind. Rather than needing to constantly watch an IPFS node to know if it's in a GC, I figured we could actually run the GC process from cluster itself.

@Stebalien do you know if there's a way that cluster could easily "hook in" and orchestrate a GC on an IPFS node?

**Comment by obo20** [comment]
Another thing to keep in mind here is that we probably want to keep the cluster API itself working on the nodes as they go through GC. 

From what I understand (correct me if I'm wrong @hsanjuan ) - Cluster already routes requests to other nodes if the node the request came on is full. Theoretically this would work the same with GC (if the node is in GC, the requests can still come in validly, it just gets rerouted to the other nodes who aren't in GC)

**Comment by lanzafame** [comment]
@kishansagathiya has a PR open for adding functionality to perfom GCs on ipfs nodes using the `/repo/gc` api endpoint. I am not sure of what it's current state is though.

> Cluster already routes requests to other nodes if the node the request came on is full

Not quite, the peer that receives the request determines which peers will actually pin the file via the allocator component (default is a sorted list of peers based on freespace and pick the top n, where n is the replication factor), basically, the peer that receives the request doesn't necessarily pin the file regardless of it's current state (i.e. full).

**Comment by obo20** [comment]
@lanzafame Good to know. That's an important distinction. So in terms of GC, this would mean that we would need to have the allocator component on each node updated when a node enters GC so that nodes in GC aren't returned from that list?

**Comment by lanzafame** [comment]
The `Informer` component, technically, is what would get and propagate the status of whether a node was GCing. The `Allocator` is what then takes that information and provides the sorted list of which peers to pin to. Though, until IPFS exposes the current status of whether the node is GCing, cluster would have to base it off whether it's request to `/repo/gc` had returned. This would "work" but this would only be good for GCs that Cluster has triggered and not for ones that the node itself has triggered.

**Comment by obo20** [comment]
That would be more than fine by me for right now (having cluster manager the GC and wait for the return). @Stebalien do you see anything wrong with that approach?

My only concern would be timeouts. I just timed a garbage collection on a copy I made of one of our storage nodes. It took around 8 hours to GC around 600GB of data on a badgerDB (SSD)

**Comment by hsanjuan** [comment]
> My only concern would be timeouts. I just timed a garbage collection on a copy I made of one of our storage nodes. It took around 8 hours to GC around 600GB of data on a badgerDB (SSD)

ouff

**Comment by RubenKelevra** [comment]
> The `Informer` component, technically, is what would get and propagate the status of whether a node was GCing. The `Allocator` is what then takes that information and provides the sorted list of which peers to pin to. Though, until IPFS exposes the current status of whether the node is GCing, cluster would have to base it off whether it's request to `/repo/gc` had returned. This would "work" but this would only be good for GCs that Cluster has triggered and not for ones that the node itself has triggered.

Is there a way to have the ipfs-cluster get notified when GC is about to start, that it can send a message in the cluster, to suspend pin operations?

If not, isn't there just a threshold for GC in the ipfs setting? The cluster could read it and trigger the GC before the threshold is reached.

This would help to avoid pinning new stuff to the node (if the cluster isn't a full copy on every node type).

Is there a way to determine how the progress for GC is doing?

It would be surely helpful for the 'Allocator' to know which node is going to finish the GC first - if there are multiple nodes in GC at the same time. This way the cluster can recover from a below minimum replication state faster, because the first cluster member to expect coming out of garbage collection can pickup the missing pieces.

While running a GC we could just write down all pi /unpin/update/add operations in a simple list form, to avoid running any commands on the IPFS-API until the GC is finished. Afterwards queue is just played back to catch up with the cluster.

**Comment by hsanjuan** [comment]
> While running a GC we could just write down all pi /unpin/update/add operations in a simple list form, to avoid running any commands on the IPFS-API until the GC is finished. Afterwards queue is just played back to catch up with the cluster.

I think IPFS will simply lock the Pinner, so essentially things will queue when you are GC-ing (and that's why it takes longer).

> Is there a way to determine how the progress for GC is doing?

No, there is no way to know how much to GC there is and no total progress.

Given that cluster will just call GC on one peer at a time and that pinning is async etc, the cluster should keep working relatively OK even if one of the peers is down to a crawl.

**Comment by RubenKelevra** [comment]
True, I've tested the GC performance on a single spinning desktop harddrive, running flatfs on ZFS (with sync in ZFS turned off).

Performance was just slightly impacted, I was pleasantly surprised. :)
